### PR TITLE
Support ping pong

### PR DESCRIPTION
**Pull request by MohamedElmdary**

### CODE REVIEW

File: lib/client.ts [hunk 1]
```diff
@@ -50,6 +49,17 @@ class Client {
...
 
+    private async __pingPong() {
+        const reqId = await this.ping();
+        return this
+            .read(reqId)
+            .catch(() => null)
+            .finally(() => {
+                if (this.con?.readyState === this.con.OPEN)
+                    this.__pingPong()
+            });
```
Comment (amiraabouhadid) [inline lib/client.ts:64]: i don't see where you keep sending ping every 20s and where you reconnect if no pong is received after 40s

File: lib/client.ts [hunk 2]
```diff
@@ -276,6 +331,9 @@ class Client {
                         this.responses.delete(requestID);
                         reject(`${err.code} ${err.message}`);
                     }
+                } else if (envelope && envelope.pong) {
+                    resolve(envelope.pong)
```
Comment (amiraabouhadid) [inline lib/client.ts:360]: needs to retry reading for pong for 40s if no pong is received

**Review by amiraabouhadid — CHANGES_REQUESTED** [review]

File: lib/client.ts (hunk 2 above)
Comment (MohamedElmdary) [reply lib/client.ts:360]: Currently if no pong is received within 40s it should recall ping method again

File: lib/client.ts [hunk 3]
```diff
@@ -50,6 +49,21 @@ class Client {
...
 
+    private __pingPongTimeout?: NodeJS.Timeout;
+    private async __pingPong() {
+        if (this.__pingPongTimeout) clearTimeout(this.__pingPongTimeout);
+        const reqId = await this.ping();
+        return this
+            .read(reqId)
+            .catch(() => this.reconnect())
+            .finally(() => {
+                this.__pingPongTimeout = setTimeout(() => {
```
Comment (amiraabouhadid) [inline lib/client.ts:60]: shouldn't this be setInterval? setTimout is executed only once, we want it executing every 20 seconds

File: lib/client.ts (hunk 2 above)
Comment (amiraabouhadid) [reply lib/client.ts:360]: I don't see the 40 s anywhere, where is it please

**Review by amiraabouhadid — CHANGES_REQUESTED** [review]

**Review by amiraabouhadid — APPROVED** [review]

**Pull request merged**